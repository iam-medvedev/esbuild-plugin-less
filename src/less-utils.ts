import fs from 'fs';
import path from 'path';
import { PartialMessage } from 'esbuild';

const importRegex = /@import.*?["']([^"']+)["'].*?/;
const globalImportRegex = /@import.*?["']([^"']+)["'].*?/g;
const importCommentRegex = /(?:\/\*(?:[\s\S]*?)\*\/)|(\/\/(?:.*)$)/gm;

const extWhitelist = ['.css', '.less'];

/** Recursively get .less/.css imports from file */
export const getLessImports = (filePath: string): string[] => {
  try {
    const dir = path.dirname(filePath);
    const content = fs.readFileSync(filePath).toString('utf8');

    const cleanContent = content.replace(importCommentRegex, '');
    const match = cleanContent.match(globalImportRegex) || [];

    const fileImports = match
      .map((el) => {
        const match = el.match(importRegex);
        return match[1];
      })
      .filter((el) => !!el)
      // NOTE: According to the docs, extensionless imports are interpreted as '.less' files.
      // http://lesscss.org/features/#import-atrules-feature-file-extensions
      // https://github.com/iam-medvedev/esbuild-plugin-less/issues/13
      .map((el) => path.resolve(dir, path.extname(el) ? el : `${el}.less`));

    const recursiveImports = fileImports.reduce((result, el) => {
      return [...result, ...getLessImports(el)];
    }, fileImports);

    const result = recursiveImports.filter((el) => extWhitelist.includes(path.extname(el).toLowerCase()));

    return result;
  } catch (e) {
    return [];
  }
};

/** Convert less error into esbuild error */
export const convertLessError = (error: Less.RenderError): PartialMessage => {
  const sourceLine = error.extract.filter((line) => line);
  const lineText = sourceLine.length === 3 ? sourceLine[1] : sourceLine[0];

  return {
    text: error.message,
    location: {
      namespace: 'file',
      file: error.filename,
      line: error.line,
      column: error.column,
      lineText,
    },
  };
};
