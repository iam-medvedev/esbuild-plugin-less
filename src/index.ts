import type { Loader, Plugin } from 'esbuild';
import path from 'path';
import { promises as fs } from 'fs';
import less from 'less';
import { convertLessError, getLessImports } from './less-utils';

export interface LoaderOptions {
  /* custom filter */
  filter?: RegExp;
}

/** Less-loader for esbuild */
export const lessLoader = (options: Less.Options = {}, loaderOptions: LoaderOptions = {}): Plugin => {
  return {
    name: 'less-loader',
    setup: (build) => {
      const filter = loaderOptions.filter;

      // Resolve *.less files with namespace
      build.onResolve({ filter: filter || /\.less$/, namespace: 'file' }, async (args) => {
        const pathResolve = await build.resolve(args.path, {
          kind: args.kind,
          importer: args.importer,
          resolveDir: args.resolveDir,
          pluginData: args.pluginData,
        });
        const filePath = pathResolve.path;

        return {
          path: filePath,
          watchFiles: [filePath, ...getLessImports(filePath, options.paths || [])],
        };
      });

      // Build .less files
      build.onLoad({ filter: filter || /\.less$/, namespace: 'file' }, async (args) => {
        const content = await fs.readFile(args.path, 'utf-8');
        const dir = path.dirname(args.path);
        const basename = path.basename(args.path);
        const isModule = basename.endsWith('.module.less');
        const loader: Loader = isModule ? 'local-css' : 'css';

        const opts: Less.Options = {
          filename: args.path,
          relativeUrls: true,
          ...options,
          paths: [...(options.paths || []), dir],
        } as any;

        try {
          const result = await less.render(content, opts);

          return {
            contents: result.css,
            loader,
            resolveDir: dir,
          };
        } catch (e) {
          return {
            errors: [convertLessError(e)],
            resolveDir: dir,
          };
        }
      });
    },
  };
};
