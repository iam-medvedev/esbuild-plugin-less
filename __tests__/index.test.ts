import { describe, expect, it } from 'vitest';
import * as path from 'path';
import { build, BuildOptions } from 'esbuild';
import { lessLoader, LoaderOptions } from '../src/index';

const entryPoints = [
  path.resolve(__dirname, '../', 'example', 'index.ts'),
  path.resolve(__dirname, '../', 'example', 'index-custom-filter.ts'),
  path.resolve(__dirname, '../', 'example', 'index.less'),
  path.resolve(__dirname, '../', 'example', 'test.module.ts'),
];

const commonOptions: BuildOptions = {
  bundle: true,
  write: false,
  minify: true,
  outdir: path.resolve(__dirname, 'output'),
  loader: {
    '.ts': 'ts',
    '.png': 'dataurl',
    '.jpg': 'dataurl',
  },
};

type BuildLessProps = {
  lessOptions?: Less.Options;
  loaderOptions?: LoaderOptions;
  entryPoint?: string;
  buildOptions?: BuildOptions;
};

async function buildLess({
  lessOptions,
  loaderOptions,
  entryPoint = entryPoints[0],
  buildOptions: _buildOptions = {},
}: BuildLessProps = {}) {
  const buildOptions: BuildOptions = {
    ...commonOptions,
    entryPoints: [entryPoint],
    plugins: [lessLoader(lessOptions, loaderOptions)],
    ..._buildOptions,
  };

  const { outputFiles } = await build(buildOptions);
  return outputFiles;
}

describe('less-loader', () => {
  it('exported module', () => {
    expect(lessLoader).toBeDefined();
  });

  it('builds successful', async () => {
    const primaryColor = '#ff0000';
    const result = await buildLess({
      lessOptions: {
        globalVars: {
          primaryColor,
        },
      },
    });

    // Result has compiled .less
    expect(result![1].text).toMatchSnapshot();
  });

  it('builds successful custom filter', async () => {
    const primaryColor = '#ff0000';
    const result = await buildLess({
      lessOptions: {
        globalVars: {
          primaryColor,
        },
      },
      loaderOptions: {
        filter: /\._?less_?$/,
      },
      entryPoint: entryPoints[1],
    });

    // Result has compiled .less
    expect(result![1].text).toMatchSnapshot();
  });

  it('builds successful with less as entrypoint', async () => {
    const primaryColor = '#ff0000';
    const result = await buildLess({
      lessOptions: {
        globalVars: {
          primaryColor,
        },
      },
      entryPoint: entryPoints[2],
    });

    // Result has compiled .less
    expect(result![0].text).toMatchSnapshot();
  });

  it('builds imported .less files', async () => {
    const result = await buildLess({
      lessOptions: {
        globalVars: {
          primaryColor: '#ff0000',
        },
      },
    });

    expect(result![1].text).toMatchSnapshot();
  });

  it('builds imported ._less_ files', async () => {
    const result = await buildLess({
      lessOptions: {
        globalVars: {
          primaryColor: '#ff0000',
        },
      },
      loaderOptions: {
        filter: /\._?less_?$/,
      },
      entryPoint: entryPoints[1],
    });

    expect(result![1].text).toMatchSnapshot();
  });

  it('builds imported .css files', async () => {
    const result = await buildLess({
      lessOptions: {
        globalVars: {
          primaryColor: '#ff0000',
        },
      },
    });

    expect(result![1].text).toMatchSnapshot();
  });

  it('catches less error', async () => {
    await expect(
      buildLess({
        lessOptions: {
          globalVars: {},
        },
      }),
    ).rejects.toThrow();
  });

  it('works with module.less', async () => {
    const result = await buildLess({
      entryPoint: entryPoints[3],
      buildOptions: {
        format: 'iife',
      },
    });

    expect(result!.length).toEqual(2);

    // Result has compiled .less
    expect(result![1].text).toMatchSnapshot();
  });
});
