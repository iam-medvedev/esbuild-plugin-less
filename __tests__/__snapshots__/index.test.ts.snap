// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`less-loader > builds imported ._less_ files 1`] = `
"#style-5-css{position:-webkit-sticky;position:sticky;background:url(data:image/png;base64,iVBORw0KGgoAAAAASURBVA==) no-repeat}#style-4-css{display:none}.style-3-less{color:red}.style-2-less{color:red;background:url(data:image/jpeg;base64,/9j/) no-repeat}body{background:#ff0000}body article{width:100px}body article:first-child{width:200px}
"
`;

exports[`less-loader > builds imported .css files 1`] = `
"#style-5-css{position:-webkit-sticky;position:sticky;background:url(data:image/png;base64,iVBORw0KGgoAAAAASURBVA==) no-repeat}#style-4-css{display:none}.style-3-less{color:red}.style-2-less{color:red;background:url(data:image/jpeg;base64,/9j/) no-repeat}body{background:#ff0000}body article{width:100px}body article:first-child{width:200px}
"
`;

exports[`less-loader > builds imported .less files 1`] = `
"#style-5-css{position:-webkit-sticky;position:sticky;background:url(data:image/png;base64,iVBORw0KGgoAAAAASURBVA==) no-repeat}#style-4-css{display:none}.style-3-less{color:red}.style-2-less{color:red;background:url(data:image/jpeg;base64,/9j/) no-repeat}body{background:#ff0000}body article{width:100px}body article:first-child{width:200px}
"
`;

exports[`less-loader > builds successful 1`] = `
"#style-5-css{position:-webkit-sticky;position:sticky;background:url(data:image/png;base64,iVBORw0KGgoAAAAASURBVA==) no-repeat}#style-4-css{display:none}.style-3-less{color:red}.style-2-less{color:red;background:url(data:image/jpeg;base64,/9j/) no-repeat}body{background:#ff0000}body article{width:100px}body article:first-child{width:200px}
"
`;

exports[`less-loader > builds successful custom filter 1`] = `
"#style-5-css{position:-webkit-sticky;position:sticky;background:url(data:image/png;base64,iVBORw0KGgoAAAAASURBVA==) no-repeat}#style-4-css{display:none}.style-3-less{color:red}.style-2-less{color:red;background:url(data:image/jpeg;base64,/9j/) no-repeat}body{background:#ff0000}body article{width:100px}body article:first-child{width:200px}
"
`;

exports[`less-loader > builds successful with less as entrypoint 1`] = `
"#style-5-css{position:-webkit-sticky;position:sticky;background:url(data:image/png;base64,iVBORw0KGgoAAAAASURBVA==) no-repeat}#style-4-css{display:none}.style-3-less{color:red}.style-2-less{color:red;background:url(data:image/jpeg;base64,/9j/) no-repeat}body{background:#ff0000}body article{width:100px}body article:first-child{width:200px}
"
`;

exports[`less-loader > works with module.less 1`] = `
".o{display:block}.o{color:#000}.button{color:gray}.foo .l{color:red}.foo .l{color:green}
"
`;
